include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h

```diff
@@ -0,0 +1,687 @@
+// -*- C++ -*-
+//===-- esimd_radix_sort_kernels.h --------------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===------------------------------------------------------===//
+
+#ifndef _ONEDPL_KT_ESIMD_RADIX_SORT_KERNELS_H
+#define _ONEDPL_KT_ESIMD_RADIX_SORT_KERNELS_H
+
+#include "oneapi/dpl/pstl/onedpl_config.h"
+
+#include <ext/intel/esimd.hpp>
+#if __has_include(<sycl/sycl.hpp>)
+#    include <sycl/sycl.hpp>
+#else
+#    include <CL/sycl.hpp>
+#endif
+
+#include <cstdint>
+#include <type_traits>
+
+#include "esimd_radix_sort_utils.h"
+#include "../../../pstl/utils.h"
+
+namespace oneapi::dpl::experimental::kt::esimd::__impl
+{
+
+template <bool __is_ascending, ::std::uint8_t __radix_bits, ::std::uint16_t __data_per_work_item,
+          ::std::uint16_t __work_group_size, typename _KeyT, typename _RngPack>
+_ONEDPL_ESIMD_INLINE void
+__one_wg_kernel(sycl::nd_item<1> __idx, ::std::uint32_t __n, _RngPack&& __rng_pack)
+{
+    using _BinT = ::std::uint16_t;
+    using _HistT = ::std::uint16_t;
+    using _DeviceAddrT = ::std::uint32_t;
+
+    constexpr ::std::uint32_t __bin_count = 1 << __radix_bits;
+    constexpr ::std::uint32_t __bit_count = sizeof(_KeyT) * 8;
+    constexpr ::std::uint32_t __stage_count = oneapi::dpl::__internal::__dpl_ceiling_div(__bit_count, __radix_bits);
+    constexpr _BinT __mask = __bin_count - 1;
+    constexpr ::std::uint32_t __hist_stride = sizeof(_HistT) * __bin_count;
+
+    constexpr ::std::uint32_t __reorder_slm_size = __data_per_work_item * sizeof(_KeyT) * __work_group_size;
+    constexpr ::std::uint32_t __bin_hist_slm_size = __hist_stride * __work_group_size;
+    constexpr ::std::uint32_t __incoming_offset_slm_size = (__bin_count + 1) * sizeof(_HistT);
+
+    // max SLM is 256 * 4 * 64 + 256 * 2 * 64 + 257*2, 97KB, when  __data_per_work_item = 256, __bin_count = 256
+    // to support 512 processing size, we can use all SLM as reorder buffer with cost of more barrier
+    __dpl_esimd_ns::slm_init<::std::max(__reorder_slm_size, __bin_hist_slm_size + __incoming_offset_slm_size)>();
+
+    const ::std::uint32_t __local_tid = __idx.get_local_linear_id();
+
+    const ::std::uint32_t __slm_reorder_this_thread = __local_tid * __data_per_work_item * sizeof(_KeyT);
+    const ::std::uint32_t __slm_bin_hist_this_thread = __local_tid * __hist_stride;
+
+    __dpl_esimd_ns::simd<_HistT, __bin_count> __bin_offset;
+    __dpl_esimd_ns::simd<_DeviceAddrT, __data_per_work_item> __write_addr;
+    __dpl_esimd_ns::simd<_KeyT, __data_per_work_item> __keys;
+    __dpl_esimd_ns::simd<_BinT, __data_per_work_item> __bins;
+    __dpl_esimd_ns::simd<_DeviceAddrT, __data_per_step> __lane_id(0, 1);
+
+    const _DeviceAddrT __io_offset = __data_per_work_item * __local_tid;
+
+    static_assert(__data_per_work_item % __data_per_step == 0);
+    static_assert(__bin_count % 128 == 0);
+    static_assert(__bin_count % 32 == 0);
```

Comment:
Why do we need both checks? And didn't we check before that `__radix_bits` is eight or is this a general check for once we support other values?

Reply:
That's right. These general checks might be useful when adding the support for more configurations. 

They are applied to each loop to guarantee no tails are left. To be more precise:

```
    static_assert(__data_per_work_item % __data_per_step == 0);
    static_assert(__bin_count % 128 == 0);
    static_assert(__bin_count % 32 == 0);
...
   for (::std::uint32_t __s = 0; __s < __data_per_work_item; __s += __data_per_step)
   ...
   for (::std::uint32_t __s = 0; __s < __bin_count; __s += 128)
...
   for (::std::uint32_t __i = 32; __i < __bin_count; __i += 32)
```